### PR TITLE
STITCH-1257 Integrate Embedded MongoDB on iOS SDK

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -72,4 +75,4 @@ Carthage/Build
 fastlane/report.xml
 fastlane/Preview.html
 fastlane/screenshots
-fastlane/test_output
```
Comment (edaniels) [inline .gitignore]: [typo]?

File: StitchCore-iOS/StitchCore-iOS/Core/Internal/StitchAppClientImpl.swift [hunk 2]
```diff
@@ -100,12 +100,12 @@ internal final class StitchAppClientImpl: StitchAppClient {
      * - returns: a service client whose type is determined by the `T` type parameter of the `AnyServiceClientProvider`
      *            passed in the `forProvider` parameter.
      */
-    public func serviceClient<T>(forService serviceClientProvider: AnyServiceClientProvider<T>) -> T {
-        return serviceClientProvider.client(
+    public func serviceClient<T>(forService serviceClientProvider: AnyServiceClientProvider<T>) throws -> T {
```
Comment (edaniels) [inline StitchCore-iOS/StitchCore-iOS/Core/Internal/StitchAppClientImpl.swift]: [minor] This isn't going to be fun to use if this can throw. Can we avoid this somehow?

File: StitchMDBKit/Podfile [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# Uncomment the next line to define a global platform for your project
```
Comment (edaniels) [inline StitchMDBKit/Podfile]: [request] Could we name this module something else? I'd rather it break a convention and be `LocalMongoDBService`. Is that bad?

File: StitchMDBKit/StitchMDBKit/StitchMDBKit.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
+//
+//  StitchMDBKit.h
+//  StitchMDBKit
+//
+//  Created by Jason Flax on 4/30/18.
```
Comment (edaniels) [inline StitchMDBKit/StitchMDBKit/StitchMDBKit.h]: [remove]

File: StitchMDBKit/StitchMDBKit/StitchMDBKit.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
+//
```
Comment (edaniels) [inline StitchMDBKit/StitchMDBKit/StitchMDBKit.h]: [q] Is this file necessary because of pod?

File: StitchMDBKit/StitchMDBKit/LocalMongoDBService.swift [hunk 6]
```diff
@@ -0,0 +1,168 @@
+import Foundation
```
Comment (edaniels) [inline StitchMDBKit/StitchMDBKit/LocalMongoDBService.swift]: [q] Should these import be split into 1st, 2nd, 3rd party?

File: StitchMDBKitExample/StitchMDBKitExample/ToDoListViewController.swift [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+        do {
+            // try to initialize the default stitch client
+            let client =
+                try Stitch.initializeDefaultAppClient(
+                    withConfigBuilder: StitchAppClientConfigurationBuilder.init {
+                        $0.clientAppId = "test-app"
+                    })
+            // try to get the default mongo client
+            let mongoClient = try client.serviceClient(forService: MongoDBService)
```
Comment (edaniels) [inline StitchMDBKitExample/StitchMDBKitExample/ToDoListViewController.swift]: [dope]

File: StitchMDBKitExample/StitchMDBKitExample/ToDoListViewController.swift [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+/// Name of ToDoList db
+private let toDoListDatabaseName = "todo"
+/// Name of ToDoList collection
+private let toDoListCollectionName = "items"
+/// Cell reuse id (cells that scroll out of view can be reused)
+private let cellReuseIdentifier = "ToDoItemTableViewCell"
+
+/// View Controller for the ToDo List
+class ToDoListViewController: UITableViewController {
```
Comment (edaniels) [inline StitchMDBKitExample/StitchMDBKitExample/ToDoListViewController.swift]: I'm not reviewing this code :). Adam can if he wants.

File: .gitignore (hunk 1 above)
Comment (jsflax) [reply .gitignore]: This is lower down in the file now

File: StitchMDBKit/Podfile (hunk 3 above)
Comment (jsflax) [reply StitchMDBKit/Podfile]: I think that's fine. I played around with the `Kit` naming because it is more "iOS"-y. But I think it's more important that we have unification throughout the SDKs.

File: StitchMDBKit/StitchMDBKit/StitchMDBKit.h (hunk 5 above)
Comment (jsflax) [reply StitchMDBKit/StitchMDBKit/StitchMDBKit.h]: I believe it's so people can import us from obj-c projects.

File: StitchMDBKit/StitchMDBKit/LocalMongoDBService.swift (hunk 6 above)
Comment (jsflax) [reply StitchMDBKit/StitchMDBKit/LocalMongoDBService.swift]: I do not see any style guides that give any advice here, so it's up to us. We can do it that way.

File: StitchMDBKit/StitchMDBKit/StitchMDBKit.h (hunk 5 above)
Comment (edaniels) [reply StitchMDBKit/StitchMDBKit/StitchMDBKit.h]: So is it? This is new code.

File: StitchMDBKit/StitchMDBKit/LocalMongoDBService.swift [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+private enum TrimMemoryCommand: String {
+    case mongoCommand = "trimMemory"
+    case aggressiveLevel = "aggressive"
+}
+
+/// Name of admin database to communicate commands to
+private let adminDatabaseName = "admin"
+
+/// Cached mongoClients associated to app ids
+private var localInstances = [String: MongoClient]()
```
Comment (adamchel) [inline StitchMDBKit/StitchMDBKit/LocalMongoDBService.swift]: [nit] why is this a global rather than a static member of a singleton like it is in Java?

Comment (adamchel) [reply StitchMDBKit/StitchMDBKit/LocalMongoDBService.swift]: The same could go for a lot of other stuff in this file.

File: StitchMDBKitExample/StitchMDBKitExample/ToDoListViewController.swift [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    private var toDoListCollection: MongoCollection!
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        do {
+            // try to initialize the default stitch client
+            let client =
+                try Stitch.initializeDefaultAppClient(
```
Comment (adamchel) [inline StitchMDBKitExample/StitchMDBKitExample/ToDoListViewController.swift]: [nit] not that it matters all that much for this example, but if this eventually going to be a public example, this should probably be called in `application(didFinishLaunching...` with the mongomobile init code and here should be `getDefaultAppClient` since that will probably be the best practice. If this VC gets loaded multiple times (if for example someone modified the example so that this VC isn't the top-level VC), this would crash the app.

File: StitchMDBKit/StitchMDBKit/LocalMongoDBService.swift (hunk 9 above)
Comment (jsflax) [reply StitchMDBKit/StitchMDBKit/LocalMongoDBService.swift]: Because that's not a particularly Swifty way to do it. It's more expressive if, these static members are only used within this file, to have them at the file level.

File: StitchMDBKitExample/StitchMDBKitExample/ToDoListViewController.swift (hunk 10 above)
Comment (jsflax) [reply StitchMDBKitExample/StitchMDBKitExample/ToDoListViewController.swift]: I was going by what we do in other examples. If we go with that approach, we should do the equivalent on Android as well.

**Comment by jsflax** [comment]
@edaniels @adamchel ready for another look

File: LocalMongoDBService/LocalMongoDBService/LocalMongoDBService.swift [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        objc_sync_enter(self)
+        defer { objc_sync_exit(self) }
+
+        // if client is cached, return it
+        if let client = localInstances[clientInfo.clientAppId] {
+            return client
+        }
+
+        // else, create a new client
+        let settings = MongoClientSettings(dbPath: clientInfo.dataDirectory.path)
```
Comment (edaniels) [inline LocalMongoDBService/LocalMongoDBService/LocalMongoDBService.swift]: This still needs to be configured to have the app id path concatted on like java: 
```
final String dbPath = String.format("%s/local_mongodb/0/", dataDir);
    final MongoClient client =
        MongoClients.create(MongoClientSettings.builder().dbPath(dbPath).build());
```
